server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,523 @@
+    private static final String INDEX_TYPE_NAME = "index";
+    private static final String INDEX_UUID_FIELD_NAME = "index_uuid";
+
+    private final NodeEnvironment nodeEnvironment;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    public LucenePersistedStateFactory(NodeEnvironment nodeEnvironment, NamedXContentRegistry namedXContentRegistry) {
+        this.nodeEnvironment = nodeEnvironment;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    CoordinationState.PersistedState loadPersistedState(BiFunction<Long, MetaData, ClusterState> clusterStateFromMetaData)
+        throws IOException {
+
+        final OnDiskState onDiskState = loadBestOnDiskState();
+
+        final List<MetaDataIndex> metaDataIndices = new ArrayList<>();
+        final List<Closeable> closeables = new ArrayList<>();
+        boolean success = false;
+        try {
+            for (final Path path : nodeEnvironment.nodeDataPaths()) {
+                final Directory directory = createDirectory(getMetaDataIndexPath(path, Version.CURRENT.major));
+                closeables.add(directory);
+                final IndexWriterConfig indexWriterConfig = new IndexWriterConfig(new KeywordAnalyzer());
+                final ConcurrentMergeScheduler concurrentMergeScheduler = new ConcurrentMergeScheduler();
+                concurrentMergeScheduler.setMaxMergesAndThreads(1, 1);
+                indexWriterConfig.setMergeScheduler(concurrentMergeScheduler);
+                indexWriterConfig.setCommitOnClose(false);
+                final IndexWriter indexWriter = new IndexWriter(directory, indexWriterConfig);
+                closeables.add(indexWriter);
+                metaDataIndices.add(new MetaDataIndex(directory, indexWriter));
+            }
+            success = true;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(closeables);
+            }
+        }
+
+        final ClusterState clusterState = clusterStateFromMetaData.apply(onDiskState.lastAcceptedVersion, onDiskState.metaData);
+        final LucenePersistedState lucenePersistedState
+            = new LucenePersistedState(nodeEnvironment.nodeId(), metaDataIndices, onDiskState.currentTerm, clusterState);
+        success = false;
+        try {
+            lucenePersistedState.persistInitialState();
+
+            for (final Path path : nodeEnvironment.nodeDataPaths()) {
+                assert Files.exists(getMetaDataIndexPath(path, Version.CURRENT.major - 1)) == false;
+            }
+            success = true;
+            return lucenePersistedState;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(lucenePersistedState);
+            }
+        }
+    }
+
+    // exposed for tests
+    Directory createDirectory(Path path) throws IOException {
+        return new SimpleFSDirectory(path);
```

Comment:
I added a comment in 8767e910e482102710e46cdc67a744c819d79c3e but I don't really have a good way to choose. Which alternative would you use?

Reply:
Since Lucene recommends opening with FSDirectory#open and you used a specific implementation, I was assuming that you had a reason for that. For such little data, I suspect it doesn't matter much, so SimpleFSDirectory is probably good.

Reply:
Ah yes, `FSDirectory#open` was what I was looking for, thanks. Adjusted in bc2545cf486088b85a01b93523d77d55bc0efddd.

Reply:
sorry for the back and forth but I now wonder whether NIOFSDirectory or SimpleFSDirectory would make a better default, the reasoning being that we have a setting to disable the usage of mmap (see https://github.com/elastic/elasticsearch/pull/32421 for background) so it would be weird for users that have this setting turned on to see that Elasticsearch keeps mmap-ing some files

Reply:
Yes, that's a good reason indeed. Reverted (with a comment to that effect) in e6e4d32df5e1c1ffb6f76eb53139808aaa732081.